### PR TITLE
feat: New module Issue1314 P2S VPN Gateway resource

### LINKED ISSUE #1314: [Module Proposal]: `avm/res/network/p2s-vpn-gateway`
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes.

### Bicep or Terraform?

Bicep

### Module Classification?

Resource Module

### Module Name

avm/res/network/p2s-vpn-gateway

### Module Details

Proposing creation of a new module to provision P2S VPN gateways within a Virtual WAN (VWAN) hub, using the Microsoft.Network/p2svpnGateways resource provider. This module would be a prerequisite to publishing an AVM pattern for Azure VWAN, and is currently a gap. This module would also require and depend on another module (maybe a child?) for VWAN VPN Server configurations, and would use the Microsoft.Network/vpnServerConfigurations resource provider.

### Do you want to be the owner of this module?

Yes

### Module Owner's GitHub Username (handle)

ericscheffler

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by ericscheffler**
## Description

This PR contains a new AVM resource for the Microsoft.Network/p2svpnGateways resource provider, which builds a P2S VPN Gateway within a VWAN Virtual Hub. This resource requires a Virtual WAN and Virtual Hub, as well as a VPN Server Configuration to deploy successfully.

Closes  [#1314](https://github.com/Azure/Azure-Verified-Modules/issues/1314)

## Pipeline Reference

[avm.res.network.p2s-vpn-gateway](https://github.com/ericscheffler/bicep-registry-modules/actions/workflows/avm.res.network.p2s-vpn-gateway.yml)

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Comment by jtracey93** [comment]
Hey @ericscheffler ,

Firstly, thanks for your work on this PR!

We have made some changes to the AVM CI, detailed below, which means we need you to update your fork to pull in these latest changes and re-run your tests to show they still are passing prior to approving and merging this PR, as we don't and it fails once merged the publishing of your module will fail and will be blocked going forward until the test pass again via additional PRs.

### Changes to CI That Have Been Made That You Need To Take Action On

- https://github.com/Azure/bicep-registry-modules/pull/3745
    - Adds WAF Security PSRule Checks using this custom baseline of rules ([`/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml`](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml))
    - **Action Required By You:**
        - Ingest/pull-in these changes into your PR branch and re-run your tests to ensure they pass
            - If any issues with the rules and you think it should apply, please ping/reach out to the core team

Any questions reach out to the AVM Core Team by tagging us in your PR here or internally via Teams

Thanks

Jack (AVM Core Team)

**Comment by ericscheffler** [comment]
Converted to draft, troubleshooting new issues with WAF test following recent CI/CD changes

**Comment by ericscheffler** [comment]
Updated & tests succeeding, marking as Ready for review

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:46]:
```suggestion
param virtualHubResourceId string
```
As per [ref](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr22---category-inputs---parametersvariables-for-resource-ids)

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:52]:
```suggestion
param vpnServerConfigurationResourceId string
```
As per [ref](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr22---category-inputs---parametersvariables-for-resource-ids)

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:188]: There seems something to be missing 😄

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:187]: Needs description

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:174]:
```suggestion
@export()
@description('A type to describe a static route for a vnet route')
type vnetRoutesStaticRoutesType = {
```

Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:174]: This is not used. Should be type be removed?

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:40]:
```suggestion
param propagatedLabelNames string[] = []
```

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/main.bicep:19]:
```suggestion
param propagatedRouteTableNames string[] = []
```

File: avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep:4]: Incorrect description. I guess you meant
```suggestion
metadata name = 'Using large parameter set'
metadata description = 'This instance deploys the module with most of its features enabled.'
```

File: avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep:60]:
```suggestion
```
Not needed for WAF-alignment

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:46]: Corrected all instances of that parameter name

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:52]: Corrected

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:188]: Fixed... I'm dumb

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:187]: Weird... branches may be out of sync? I see it on my end:
image

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:175]: Added description and export decorators

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:174]: It should be (used); somehow I'm missing a parameter... fixing now.

Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:174]: Added description and export decorators

File: avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep:60]: Removed

File: avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep:4]: Fixed; copy & paste error :-D

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:19]: Fixed

File: avm/res/network/p2s-vpn-gateway/main.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/main.bicep:40]: Fixed

File: avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep:50]: Please note that the following parameters are defined even though they're optional (and 'should' hence not be defined in the defaults test):
```
associatedRouteTableName: 'defaultRouteTable'
location: resourceLocation
vpnClientAddressPoolAddressPrefixes: [
  '10.0.2.0/24'
]
```

Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep:50]: For the defaults test; are we talking:

- the minimum configuration to get the resource to actually deploy
- or, the minimum configuration to deploy _and_ have it actually be functional?

Comment (AlexanderSehr) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep:50]: 😄 

The ideally both? the minimum test should ensure that the defaults of the module work (hence the name) which usually should still mean the resource works. If this is not possible for this resource type because e.g., there are different deployment variants that require different parameters to be provided maybe deploying one variant is the better idea.
At the end of the day the idea should just be to make sure the module works the way it is intended. So let's take a closer look at the 3 highlighted parameters
- `associatedRouteTableName`: I guess this could be a module default, if required? Unless this breaks something else.
- `location`: Likely not required as the module's default should anyways be the one provided
- `vpnClientAddressPoolAddressPrefixes`: The one that may have the biggest impact on functionality, I guess? And hence maybe the one to keep around.

File: avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep:18]: Beware, this serviceshort is already used by a different module (`avm\res\network\vpn-server-configuration\tests\e2e\defaults\main.test.bicep`). Is this a copy-paste error?
If not too long, I'd strongly recommend changing this short to `npvgmin`. The same comment applies to the other test files.

File: avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep:70]:
```suggestion
```
Seems redundant. Please not this removal requires a re-run of the `Set-AVMModule` script

Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep:70]: Not quite sure why I didn't remove that? My fault for working on this late from a hotel room :-)

File: avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/defaults/main.test.bicep]: Yup, copy & paste error, fixed as recommended

File: avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep:18]:
```suggestion
param serviceShort string = 'npvgmax'
```

Comment (AlexanderSehr) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep:18]: Min only applies to the defaults test

File: avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/network/p2s-vpn-gateway/tests/e2e/waf-aligned/main.test.bicep:18]:
```suggestion
param serviceShort string = 'npvgwaf'
```

`min` only applies to the defaults test

File: avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep
Comment (ericscheffler) [reply avm/res/network/p2s-vpn-gateway/tests/e2e/max/main.test.bicep:18]: Fixed, dumb copy & paste error

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**
